pythoncode/usbTrainer.py

```diff
@@ -92,477 +101,826 @@
+    # Load = Force [N] * 137 (the result of a fitting in ergo mode) one get:
+    #
+    # Load = 137 * 80 * 9.81 * slope_in_percent / 100
+    #
+    # SlopeScale = 137 * 80 * 9.81 / 100 = 1075
+    #
+    # Tests:
+    # factor 650 causes Zwift to think we ride @ 20km/h where the wheel runs at 40km/hr
+    # factor 1300 to be tested
+    #
+    #---------------------------------------------------------------------------
+    # Refer to ANT msgUnpage51_TrackResistance()
+    #          where zwift grade does not seem to match the definition
+    #---------------------------------------------------------------------------
+    # Input         TargetGrade, UserAndBikeWeight, SpeedKmh
+    #
+    # Function      Calculate what power must be produced, given the current
+    #               grade, speed and weigth
+    # 
+    # Output        TargetPower
+    #---------------------------------------------------------------------------
+    def Grade2Power(self):                                         #Grade2Power#
+        if self.TargetGrade < -20 or self.TargetGrade > 30:
+            logfile.Console("Grade2Power; TargetGrade (%s) out of bounds!" % (self.TargetGrade))
+
+        if self.UserAndBikeWeight < 70:
+            self.UserAndBikeWeight = 75 + 10
 
-    return rtn
+        self.TargetPower  = self.__PfietsNL()
 
-def PfietsNL(TargetGrade, UserAndBikeWeight, SpeedKmh):
+        if debug.on(debug.Application):
+            logfile.Write ("Grade2Power; TargetGrade=%4.1f%%, Speed=%4.1f, Weight=%3.0f, TargetPower=%3.0fW" % \
+                (self.TargetGrade, self.SpeedKmh, self.UserAndBikeWeight, self.TargetPower) )
+
+    def __PfietsNL(self):
         #-----------------------------------------------------------------------
         # Thanks to https://www.fiets.nl/2016/05/02/de-natuurkunde-van-het-fietsen/
         # Required power = roll + air + slope + mechanical
         #-----------------------------------------------------------------------
         c     = 0.004                           # roll-resistance constant
-        m     = UserAndBikeWeight               # kg
+        m     = self.UserAndBikeWeight          # kg
         g     = 9.81                            # m/s2
-        v     = SpeedKmh / 3.6                  # m/s       km/hr * 1000 / 3600
+        v     = self.SpeedKmh / 3.6             # m/s       km/hr * 1000 / 3600
         Proll = c * m * g * v                   # Watt
 
         p     = 1.205                           # air-density
         cdA   = 0.3                             # resistance factor
         w     =  0                              # wind-speed
         Pair  = 0.5 * p * cdA * (v+w)*(v+w)* v  # Watt
 
-        i     = TargetGrade
+        i     = self.TargetGrade
         Pslope= i/100 * m * g * v               # Watt
 
         Pbike = 37
 
-        return Proll + Pair + Pslope + Pbike
+        return int(Proll + Pair + Pslope + Pbike)
+
+    def __VfietsNL(self):

```

Comment:
19:36:22,091: Target=-1.7%(43W) Speed=33.8kmh hr=  0 Current= 64W Cad= 77 r= 546  29
Traceback (most recent call last):
  File "FortiusAnt.py", line 520, in <module>
    Console.Autostart()
  File "FortiusAnt.py", line 161, in Autostart
    Tacx2Dongle(self)
  File "FortiusAnt.py", line 110, in Tacx2Dongle
    rtn = FortiusAntBody.Tacx2Dongle(self)
  File "/home/pi/WouterJD/FortiusANT/pythoncode/FortiusAntBody.py", line 448, in Tacx2Dongle
    rtn = Tacx2DongleSub(self, Restart)
  File "/home/pi/WouterJD/FortiusANT/pythoncode/FortiusAntBody.py", line 684, in Tacx2DongleSub
    TacxTrainer.Refresh()               # = Receive + Calc + Send
  File "/home/pi/WouterJD/FortiusANT/pythoncode/usbTrainer.py", line 728, in Refresh
    super().Refresh(TacxMode)
  File "/home/pi/WouterJD/FortiusANT/pythoncode/usbTrainer.py", line 445, in Refresh
    self.VirtualSpeedKmh = self.__VfietsNL()
  File "/home/pi/WouterJD/FortiusANT/pythoncode/usbTrainer.py", line 597, in __VfietsNL
    v     = 3.6 * ( (P - 37) / (c * m * g + i/100 * m * g) )
ZeroDivisionError: float division by zero

Reply:
solved. Sorry crappy code.